widgetsnbextension/src/manager.js

```diff
@@ -190,12 +188,21 @@ WidgetManager.set_state_callbacks = function (load_callback, save_callback, opti
 var url = [window.location.protocol, '//', window.location.host, window.location.pathname].join('');
 var key = 'widgets:' + url;
 WidgetManager.set_state_callbacks(function() {
-    if (localStorage[key]) {
-        return JSON.parse(localStorage[key]);
+    if (Jupyter.notebook.metadata.widgets) {
+        return return Promise.resolve(Jupyter.notebook.metadata.widgets.state);
     }
-    return {};
+    return Promise.resolve({});
 }, function(state) {
-    localStorage[key] = JSON.stringify(state);
+    Jupyter.notebook.metadata.widgets = {
+
+        // Persisted widget state version
+        version: '1.0.0',
```

Comment:
I am hesitant to add a version number that we are not using from the start. The risk is that it would be stale - but I see where you come from.